README.md
# [picture2color](./)

[![npm version][npm-version-img]][npm-url]
[![npm download][npm-download-img]][npm-download]
[![MIT License][license-image]][license-url]
[![issues-img]][issues]
[![pr-img]][pr]
[![size-img]][size-img]
[![file-img]][file-img]

## 简介
此模块旨在解决前端在图片颜色识别上的问题，有问题可以联系
[我][e-mail]或者提[issues][issues]

1、能够分析图片的主要颜色

2、根据坐标(像素/百分比)获取对应颜色

3、判定单个或者多个颜色深浅判定

4、对比两个颜色是否在参数程度范围内是否相似

5、获取颜色(或范围颜色)于颜色数据的占比

6、获取图片的全部颜色数据

7、按图片边框由外到内0-0.5占比的颜色数据

8、根据颜色数据创建范围颜色代理colorProxy

![img][demo-img]

## Demo

[链接][demo]

## install
``` bash
    npm install --save picture2color
```

## import

```javascript
    import Picture2color from 'Picture2color'
```

## static API

1、是否是深色
```javascript
     /**
     * pa
     * @param {Color || ColorProxy} color
     * @param {Number} colorStep 值越小说明深色判定范围越小 默认192
     * @return Boolean
     */
    static isDeep(color, deepStep)
```
2、获取坐标位置传入图片的Color instance
```javascript
    /**
     * 获取图片坐标内色值
     * @param {*} x
     * @param {*} y
     * @param {*} image
     * @param {*} isPiex
     * @return Color
     */
    static getImageColor(x, y, image, isPiex)
```
3、判定
```javascript
    /**
     * 颜色是否在对应的范围内色值相似
     * @param {Color | [r,g,b,a] | colorProxy} color1
     * @param {Color | [r,g,b,a] | colorProxy} color2
     * @param {number 1++} colorStep 值越小范围越小,默认100
     */
    static isSimilarColor(color1, color2, colorStep)
```
4、颜色类（传入[r,g,b,a]）实例化color对象
```javascript
    static Color
```
## 实例化

1、实例化模块
```javascript
    const image = new Image() || "http://url"
    const options = {
        async: false, // 是否异步化执行 传入图片为字符串链接也会默认转化为async执行
        event: ['click'], // 绑定事件获取颜色信息 然后通过emit=>color向外反馈
        deepStep: 192 // 判定深浅色程度，值越大深浅灵敏度越小  --Color
    }
    let demo = new Picture2color(image, options);
    demo instanceof Picture2color // true
    demo.on('color', () => {});
```

## 实例化API

1、获取当前图片主要色值（数据量大的时候计算时间会比较长）

```javascript
    /**
     * 获取图片占比主要颜色列表
     */
    instance.getColorGroup() // 返回ColorGroup实例
```
2、以边框为界限向内获取主要颜色列表（数据量大的时候计算时间会比较长）
```javascript
     /**
     * 以边框为界限向内获取0-0.5范围主要颜色列表
     * @param {*} option {size: 0-0.5}
     */
    instance.getFrameColorGroup() // 返回ColorGroup实例
```

## Color

1、实例化
```javascript
    /**
     * @params {[r,g,b,a]} rgba色值数组
     * @params {*} {deepStep: 0-255} 深色值 值越小说明深色判定范围越小
     */
    let colorInstance = new Picture2color.Color(data, option)
```
2、实例化API

```javascript
    colorInstance.value // rgba 色值
    colorInstance.isDeep // Boolean
    colorInstance.count // Number
    colorInstance.percent // Number 如果是按组生成的颜色可查看百分比
```
## ColorGroup

1、实例化
```javascript
    /**
     * @params {color} 参数传入color时会将color设置为group的proxy 将color属性代理到颜色组，为空时则纯作为颜色组
     * 所谓colorProxy就是将颜色差距极小的颜色归类到一组，然后取组内占比最大的的颜色属性作为代表属性，既包含colorGroup特性也包含color属性
     */
    let groupInstance = new Picture2color.ColorGroup()
```
2、实例化API

```javascript
    groupInstance.list // Array 颜色组包含的所有颜色列表
    groupInstance.sortList // Array 颜色组包含的所有颜色列表-降序排列
    groupInstance.count // Number
    groupInstance.percent // Number 如果是按组生成的颜色可查看百分比

    const option = {colorStep: 100}
    groupInstance.createColorProxy(option) // 将colorGroup生成一个新colorGrou，colorGroup内包的的是ColorProxy
    /****如果是colorProxy将额外获得Color实例的主要属性****/

```
## Count(Color和ColorGroup 都继承的类)

1、实例化
```javascript
    /**
     */
    let countInstance = new Count()
```
2、实例化API

```javascript
    countInstance.count // number 数量
    countInstance.percent // Number 百分比
```

## 小更新
1. isSimilarColor等静态方法支持字符串#fff、rgba(255,255,255)

## 未来可期

1. 同颜色坐标查找
2. 颜色范围坐标查找
3. serviceWorker解决性能问题

[demo-img]: ./assets/color.png
[demo]: https://wanxiaodong404.github.io/picture2color/examples/

[npm-version-img]: https://img.shields.io/npm/v/picture2color
[npm-url]: https://www.npmjs.com/package/picture2color

[npm-download-img]: https://img.shields.io/npm/dw/picture2color.svg?style=flat
[npm-download]: https://npmcharts.com/compare/picture2color?minimal=true

[license-image]: https://img.shields.io/badge/license-MIT-blue.svg?style=flat
[license-url]: LICENSE

[e-mail]: mailto://729779978@qq.com

[issues-img]: https://img.shields.io/bitbucket/issues-raw/wanxiaodong404/picture2color
[issues]: https://github.com/wanxiaodong404/picture2color/issues


[size-img]: https://img.shields.io/badge/minified%20size-16%20kB-informational
[file-img]: https://img.shields.io/badge/files-36-blue
[pr-img]: https://img.shields.io/bitbucket/pr-raw/wanxiaodong404/picture2color
[pr]: https://github.com/wanxiaodong404/picture2color/pr


CHANGELOG.md
## [0.1.6](https://github.com/sheldonWan/picture2color/compare/v0.1.3...v0.1.6) (2021-09-09)


### Features

* 静态方法优化 ([e4edb5d](https://github.com/sheldonWan/picture2color/commit/e4edb5da8c174c292042bff2c0552af9b7a7005f))
* build ([91cb7c7](https://github.com/sheldonWan/picture2color/commit/91cb7c72d106f4a4d938ab44eadc5c5912302d2e))
* new version ([da94a96](https://github.com/sheldonWan/picture2color/commit/da94a96ce09c16fe1ed3028ca28b147af4210a1f))



## [0.1.3](https://github.com/sheldonWan/picture2color/compare/v0.1.2...v0.1.3) (2021-01-26)


### Bug Fixes

* import ([7240d0c](https://github.com/sheldonWan/picture2color/commit/7240d0c3b1fdd61955116e1d5829e5ffe5ceb4ae))



## [0.1.2](https://github.com/sheldonWan/picture2color/compare/v0.1.1...v0.1.2) (2021-01-26)



## [0.1.1](https://github.com/sheldonWan/picture2color/compare/v0.1.0...v0.1.1) (2020-12-23)


### Bug Fixes

* fixed count bug ([88fb927](https://github.com/sheldonWan/picture2color/commit/88fb927fd34cd2f165dcebbe962b7327bee49b6b))
* fixed count bug ([829c1dc](https://github.com/sheldonWan/picture2color/commit/829c1dcce3be8aa91811da0e325bc2fc78878147))


### Features

* keyword ([a117dbe](https://github.com/sheldonWan/picture2color/commit/a117dbe84f522205b30ad0dad0cd3bcafa08ec6d))



# [0.1.0](https://github.com/sheldonWan/picture2color/compare/v0.0.10...v0.1.0) (2020-12-23)


### Bug Fixes

* hex string bug fixed ([152ae7c](https://github.com/sheldonWan/picture2color/commit/152ae7cc80bb28a63d12248fa0263e8bc28ccd8e))


### Features

* 添加types和paramsFilter ([1540475](https://github.com/sheldonWan/picture2color/commit/1540475e64fbf679e28c6c7165a1d224e2b796f9))
* 重新规划0.1.0版本 ([2c7861f](https://github.com/sheldonWan/picture2color/commit/2c7861f87e89baec82dc5e189ddc4f581967488c))
* 重新设计color proxy ([7581427](https://github.com/sheldonWan/picture2color/commit/758142795585a103984f94dbfcc82e76a485ad54))



## [0.0.10](https://github.com/sheldonWan/picture2color/compare/v0.0.9...v0.0.10) (2020-11-27)


### Features

* 添加changelog ([fa7aa18](https://github.com/sheldonWan/picture2color/commit/fa7aa18b8c21810e5fd9c80e898591332ab93094))
* 新添加point类 ([e45f673](https://github.com/sheldonWan/picture2color/commit/e45f673712c75b4c5e2b1638c412a2fd35b74283))
* 新增异步执行参数以及changlog调整 ([a416ee9](https://github.com/sheldonWan/picture2color/commit/a416ee9429c9b9db8fd4cfd2fe37d95275fd1534))
* 新增demo ([6a89dc3](https://github.com/sheldonWan/picture2color/commit/6a89dc3afcb71bd24ad2b56e47f5c993d8c542d9))



## [0.0.9](https://github.com/sheldonWan/picture2color/compare/v0.0.7...v0.0.9) (2020-11-26)



## [0.0.7](https://github.com/sheldonWan/picture2color/compare/v0.0.6...v0.0.7) (2020-10-26)



## [0.0.6](https://github.com/sheldonWan/picture2color/compare/v0.0.5...v0.0.6) (2020-10-26)



## [0.0.5](https://github.com/sheldonWan/picture2color/compare/0.0.4...v0.0.5) (2020-10-22)



## [0.0.4](https://github.com/sheldonWan/picture2color/compare/0.0.3...0.0.4) (2020-10-22)



## 0.0.3 (2020-10-22)





src/lib/main.js
/*
 * @Author: wanxiaodong
 * @Date: 2020-10-19 16:36:09
 * @Last Modified by: wanxiaodong
 * @Last Modified time: 2021-09-09 10:12:16
 * @Description:
 */
const events = require('events')
const ColorAnalyse = require('./Analyse')
const Color = require('./Color')
const ColorGroup = require('./ColorGroup')
const utils = require('../utils')
const paramsFilter = require('../utils/paramsFilter')
const types = require('../types')
const defaultOption = {
    async: false, // 是否异步化执行 传入图片为字符串链接也会默认转化为async执行
    event: ['click'], // 绑定事件获取颜色信息 然后通过emit=>color向外反馈
    deepStep: 192 // 判定深浅色程度，值越大深浅灵敏度越小  --Color
}
class Picture2color extends events {
    constructor(image, option) {
        super();
        this.originColorData = null
        this.colorData = null;
        this.option = paramsFilter.main(Object.assign({}, defaultOption, (option || {})));
        this.__el = null;
        this.__cache = {};
        // this.utils = utils 不对外暴露
        if (this.option.async || typeof image === 'string') {
            return this.asyncInit(image)
        } else {
            if (image.complete) {
                this.init(image)
            } else {
                throw Error('图片尚未加载完成')
            }
        }
    }
    /**
     * 异步初始化
     * @param {*} image
     */
    asyncInit(image) {
        let that = this;
        return new Promise(function(resolve, reject) {
            if (typeof image === 'string') {
                let url = image;
                image = new Image();
                image.setAttribute('crossOrigin', 'anonymous');
                image.onload = () => {
                    that.init(image)
                    resolve(that)
                }
                image.onerror = () => {
                    reject(new Error('图片加载失败'))
                }
                image.src = url
            } else {
                that.init(image)
                resolve(that)
            }
        })
    }
    init(image) {
        this.__el = image
        this.getImageColor = Picture2color.getImageColor;
        this.getColorData(image)
        this.destory = Picture2color.destory;
        this.option.event && this.bindEvent(this.option.event);
    }
    /**
     * 绑定事件
     * @{Array} 事件类型
     */
    bindEvent(eventType = []) {
        let that = this;
        let event = eventType.map((type) => {
            let callback = function(event) {
                var {offsetX, offsetY} = event;
                let color = that.getImageColor(offsetX, offsetY, that.__el)
                that.emit('color', {
                    type,
                    eventPosition: [offsetX, offsetY],
                    color: color
                })
            };
            this.__el.addEventListener(type, callback)
            return {
                type,
                callback
            }
        });
        this.__cache.event = event
    }
    /**
     * 解绑事件
     */
    unbindEvent() {
        let that = this;
        if (this.__cache.event) {
            this.__cache.event.forEach(({type, callback}) => {
                that.__el.removeEventListener(type, callback)
            })
        }
    }
    /**
     * 获取图片的颜色数据
     * @param {*} image
     */
    getColorData(image) {
        let canvas = document.createElement('canvas');
        let {width, height} = image;
        canvas.width = width
        canvas.height = height
        this.__canvas = canvas
        let ctx = canvas.getContext('2d')
        ctx.drawImage(image, 0, 0);
        let data = ctx.getImageData(0, 0, width, height);
        this.originColorData = data;
        this.analyData = new ColorAnalyse(data, utils.paramsFilter.analy(this.option));
    }
    /**
     * 颜色筛选  性能问题 暂时不对外开放
     * @param {Color | [rgba]} color
     */
    colorFilter_bate(color) {
        return this.analyData.colorFilter_bate(color)
    }
    /**
     * 获取图片主要颜色列表
     */
    getColorGroup() {
        return this.analyData.colorGroup
    }
    /**
     * 以边框为界限向内获取0-0.5范围主要颜色列表
     * @param {*} option {size: 0-0.5}
     */
    getFrameColorGroup(option) {
        return this.analyData.getFrameColorGroup(option);
    }
    /**
     * 判断颜色或者颜色列表是否是深色
     * @param {Color || ColorList} colorList
     * @param {Number} deepStep
     */
    static isDeep(colorList, deepStep) {
        if (colorList instanceof Color) {
            return deepStep ? utils.isDeep(colorList, deepStep) : colorList.isDeep;
        }
        return colorList.reduce((percent, color) => {
            let isDeep = deepStep ? utils.isDeep(color, deepStep) : color.isDeep;
            return percent + (isDeep ? color.percent : 0)
        }, 0) > 50
    }
    /**
     * 销毁实例
     * @param {*} instance
     */
    static destory(instance) {
        instance = instance || this
        instance.unbindEvent()
        Object.keys(instance.__cache).forEach(key => {
            delete instance.__cache[key]
        })
        instance.__cache = null;
        instance.originColorData = null;
        instance.option = null;
        instance.__el = null;
        instance.off('.inner');
        instance.analyData && instance.colorData.destory();
        instance.colorData = null;
    }
    /**
     * 获取图片坐标内色值
     * @param {*} x
     * @param {*} y
     * @param {*} image
     * @param {*} isPiex
     */
    static getImageColor(x = 0, y = 0, image, isPiex = true) {
        try {
            if (this instanceof Picture2color) {
                let {width, height, naturalHeight, naturalWidth} = image;
                if (!isPiex) {
                    // 百分比模式
                    x = Math.round(x * width * 0.01);
                    y = Math.round(y * height * 0.01);
                } else {
                    x = Math.round(x / width * naturalWidth);
                    y = Math.round(y / height * naturalHeight);
                }
                return this.analyData.getPositionColor(x, y)
            } else {
                let {width, height, naturalHeight, naturalWidth} = image;
                let canvas = document.createElement('canvas');
                canvas.width = width
                canvas.height = height
                let ctx = canvas.getContext('2d')
                ctx.drawImage(image, 0, 0, naturalWidth, naturalHeight, 0, 0, width, height);
                if (!isPiex) {
                    // 百分比模式
                    x = Math.round(x * width * 0.01);
                    y = Math.round(y * height * 0.01);
                }
                let data = ctx.getImageData(x, y, 1, 1);
                let color = new Color(data.data)
                return color
            }
        } catch (e) {
            console.warn('获取数据失败')
            return {}
        }
    }
    /**
     * 颜色是否在对应的范围内色值相似
     * @param {Color | [r,g,b,a] ｜ string} color1
     * @param {Color | [r,g,b,a] ｜ string} color2
     * @param {number 1-255} step
     */
    static isSimilarColor(color1, color2) {
        return utils.isSimilarColor(...arguments)
    }
}

Picture2color.Color = Color
Picture2color.ColorGroup = ColorGroup

module.exports = Picture2color